zend_abstract_interface/json/json.h

```diff
@@ -0,0 +1,44 @@
+#ifndef ZAI_JSON_H
+#define ZAI_JSON_H
+
+#include "php.h"
+
+#define php_json_encode _dd_json_encode
+#define php_json_decode_ex _dd_json_decode_ex
+
+#define PHP_JSON_OBJECT_AS_ARRAY (1 << 0)
+
+#if PHP_VERSION_ID < 70000
+#include "ext/standard/php_smart_str.h"
+
+extern void (*php_json_encode)(smart_str *buf, zval *val, int options TSRMLS_DC);
+extern void (*php_json_decode_ex)(zval *return_value, char *str, int str_len, int options, long depth TSRMLS_DC);
+
+static inline void php_json_decode(zval *return_value, char *str, int str_len, zend_bool assoc, long depth TSRMLS_DC) {
+    php_json_decode_ex(return_value, str, str_len, assoc ? PHP_JSON_OBJECT_AS_ARRAY : 0, depth TSRMLS_CC);
+}
+#define smart_str_length(str) (str)->len
+#define smart_str_value(str) ((char *)(str)->c)
+#else
+#include <stdbool.h>
+
+#include "zend_smart_str.h"
+
+extern int (*php_json_encode)(smart_str *buf, zval *val, int options);
+extern int (*php_json_decode_ex)(zval *return_value, const char *str, size_t str_len, zend_long options,
+                                 zend_long depth);
+
+static inline int php_json_decode(zval *return_value, const char *str, int str_len, bool assoc, zend_long depth) {
+    return php_json_decode_ex(return_value, str, str_len, assoc ? PHP_JSON_OBJECT_AS_ARRAY : 0, depth);
+}
+#define smart_str_length(str) (str)->s->len
+#define smart_str_value(str) ((char *)(str)->s->val)
```

Comment:
Let's not define such a generic macro in unrelated header files. These are just needed for the test, right? Then these defs should also live in the test.

Reply:
They are needed for the test, but also needed for anything that uses the zai_json interface, so defining them in the test doesn't make much sense in that light, we don't want the consumer of the api to have to define them in their code.

I'm not sure why the zend_long undef is even there, it's copy pasta ?

Reply:
Not really. On PHP 7/8 you typically are just going to use ZSTR_LEN(str->s) - this would only matter for version independent access - and in general these macros fit better in compatibilty.h if needed. Let's put them in tests only for now.

Reply:
zend_long undef is leftover by me, shouldn't be there anymore.

Reply:
Okay, I think they fit in compat header, but the comment in compat header is confusing. It says they should only be used in headers or tests, but call sites (consumer of abstract api) need also to use them.

Maybe we should have an abstract interface for smart_str, but what do we gain from an abstract interface to json if around the callsite the consumer is forced to ifdef anyway ?

I think the interface needs to include everything you reasonably need to consume it.

Reply:
Possibly we should, once the need arises. For current non-test consumers there's no need.